---
title: "ASV correlations"
author: "Julin Maloof"
date: "5/19/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r}
library(tidyverse)
library(phyloseq)
library(magrittr)
library(knitr)
library(igraph)
library(plyr)
library(tidyverse)
library(JuliaCall)
library(UpSetR)
julia_setup()
julia_install_package_if_needed("StatsBase")
julia_library("StatsBase")
julia_library("LinearAlgebra")
```

## get ASV (from Scott Klasek)
```{r}
ps<- readRDS(file="../input/rhizo.ps")
```


```{r}
asv_counts <- otu_table(ps)# the read count data
dim(asv_counts)
head(asv_counts[,1:10])
```


```{r}
taxa <- tax_table(ps) # taxonomy data
dim(taxa)
head(taxa)
```


```{r}
asv_samples <- sample_data(ps) %>% # sample information
  as_tibble() %>%
  mutate(ID=rownames(sample_data(ps)),
         days_grown=str_pad(days_grown, 2, pad="0"), # get things formatted the same as the RNA names
         block=str_c("blk", block),
         inoc={str_replace(inoc, "Autoclaved", "ATM_BLANK") %>%
             str_replace("Inoculated", "SBC_OLD")},
         timeofday={
           str_replace(timeofday, "9:00 AM", "1_morn") %>%
             str_replace("14:00 PM", "2_afternoon") %>%
             str_replace("18:00 PM", "3_evening_5.30") %>%
             str_replace("22:00 PM", "4_night_1") %>%
             str_replace("4:00 AM", "5_night_2")},
         sampleID=str_c(inoc,days_grown,timeofday,block, sep = "-")) %>%
  select("ID", everything())
dim(asv_samples)
asv_samples
```

## get annotation
```{r}
annotation <- read_csv("../../RNAseq/Annotation_copy/output/v3.0annotation/Brapa_V3.0_annotated.csv") %>%
  select(-X1)
annotation
```

## get RNA

```{r}
cpm <- read_tsv("../../RNAseq/output/log2cpmSample.txt.gz")
head(cpm[,1:10])
```

### wrangle it

transpose and fix rownames
```{r}
rna <- cpm %>% 
  select(-gene) %>% 
  (t) %>%
  set_colnames(cpm$gene)

rownames(rna) <- str_replace(rownames(rna), "_blk", "-blk")
rna[1:5,1:2]
```

make a sample info sheet

```{r}
rna_samples <- tibble(sample=rownames(rna)) %>%
  mutate(sampleID = str_remove(sample, "^[a-zA-Z]*-")) %>%
  separate(sample, sep="-", into = c("genotype", "inoc", "days_grown", "timeofday", "block"), remove = FALSE) %>%
  select(-sample, sampleID, everything())
rna_samples
```
## get ASV and RNA order to match

```{r}
samples_merged <- inner_join(asv_samples, rna_samples) %>%
  select(ASV_ID=ID, RNA_ID=sample, sampleID, everything())
samples_merged
```


```{r}
rna <- rna[samples_merged$RNA_ID,]
head(rna[,1:10])
asv_counts <- asv_counts[samples_merged$ASV_ID,]
head(asv_counts[,1:10])
```

Let's give them the same rownames
```{r}
all(rownames(rna)==samples_merged$RNA_ID) # double check!
rownames(rna) <- samples_merged$ASV_ID
```


## filter ASVs
```{r}
sparsity <- apply(asv_counts, 2, function(x) sum(x==0)/length(x))
hist(sparsity)
sparsity[sparsity>0.9] %>% hist()
```

```{r}
asv_counts_small <- asv_counts[,sparsity< .8]
dim(asv_counts_small)
```

```{r}
asv_live <- asv_counts[asv_samples$inoc=="SBC_OLD",]
sparsity_live <- apply(asv_live, 2, function(x) sum(x==0)/length(x))
hist(sparsity_live)
sparsity_live[sparsity_live>0.9] %>% hist()
```

```{r}
asv_live_small <- asv_live[,sparsity_live< .8]
dim(asv_live_small)
```


## functions

MR matrix creation functions
```{r}
subsetRNA <- function(rna, cutoff=10000) { 
  if(ncol(rna) > nrow(rna)) { # genes in columns
    cv <- apply(rna, 2, function(x) sd(x)/abs(mean(x)))
    cutoff <- sort(cv, decreasing = TRUE)[cutoff]
    rna <- rna %>% magrittr::extract(,cv>=cutoff)
  } else { # genes in rows
    cv <- apply(rna[,-1], 1, function(x) sd(x)/abs(mean(x)))
    cutoff <- sort(cv, decreasing = TRUE)[cutoff]
    rna <- rna %>% filter(cv>=cutoff)
  }
  return(rna)
}

julia_command("myrank = function(x)
                  ordinalrank( abs.(x); rev=true)
              end")

juliaCorMR <- function(combined) {
  julia_assign("combined", as.matrix(combined))
  julia_command("cormat = cor(combined)")
  cormat <- julia_eval("cormat")
  dimnames(cormat) <- list(colnames(combined), colnames(combined))
  
  julia_command("cormat[diagind(cormat)] .= 0") # set diagonal to 0 so we don't count self ID in MR
  julia_command("rankmat = mapslices(myrank, cormat; dims=1)")
  MR <- julia_eval("sqrt.(rankmat .* transpose(rankmat))")
  dimnames(MR) <- list(colnames(cormat), colnames(cormat))
  
  return(list(cormat=cormat, MR=MR))
}

# The main function for MR matrix
getcorMR <- function(rna, asv) {
  rna <- subsetRNA(rna)
  try(if(!all(rownames(asv)==rownames(rna))) stop("row names of rna and asv don't match"))
  combined <- cbind(asv,rna)
  rm(rna)
  gc()
  return(juliaCorMR(combined))
}
```

graph functions
```{r mr_subgraphs, echo=FALSE, cache.lazy=FALSE, warning=FALSE, eval=TRUE }
get.mr.subgraph <- function(mr.cutoff,mr.matrix,annotation=NULL,neighborhood,order=1) {
  #function to extract graph at a specificed correlation cutoff
  gene.mr.tmp <- mr.matrix
  gene.mr.tmp[abs(gene.mr.tmp) > mr.cutoff] <- 0
  gene.mr.tmp[is.na(gene.mr.tmp)] <- 0 #important! otherwise vertices with NA edges are connected
  
  gene.graph <- graph.adjacency(adjmatrix = gene.mr.tmp,
                                mode="undirected",
                                weighted="mr",
                                diag=FALSE)
  
  sub.graphs <- graph.neighborhood(gene.graph,order=order,nodes=neighborhood) #for each node of interest get all other nodes within order of order
  
  #get combined list of vertices...
  
  sub.vertices <- unique(names(unlist(lapply(sub.graphs, V))))
  
  combined.sub.graph <- induced_subgraph(gene.graph,sub.vertices)
  
  V(combined.sub.graph)$color <- "lightblue"
  V(combined.sub.graph)[neighborhood]$color <- "red"
  if(!is.null(annotation)) V(combined.sub.graph)$gene <- 
    annotation$At_symbol[match(V(combined.sub.graph)$name,annotation$name)]
  return(combined.sub.graph)
}
```

plotting functions
```{r plotfunctions, echo=FALSE, cache.lazy=FALSE, eval=TRUE }
annotateEdges <- function(graph, cormat, edge.colors=c("magenta","green")) {#color edges by correlation sign and set width

  E(graph)$cor  <-  E(graph) %>% 
    attr("vnames") %>% 
    tibble(edge=.) %>% 
    separate(edge,into=c("node1","node2"),remove = FALSE,sep="\\|") %>%
    rowwise() %>% 
    mutate(cor=map2_dbl(node1, node2, ~ cormat[.x,.y])) %>%
    select(cor) %>%
    unlist()
  
  E(graph)$color <- ifelse(E(graph)$cor> 0, edge.colors[1], edge.colors[2])
  
  E(graph)$width <- rank(E(graph)$mr)^(1/4)
  
  return(graph)
}

plot.graph <- function(graph, title) {
  plot(graph,
       layout = layout_with_kk, 
       vertex.label = ifelse(is.na(V(graph)$gene)|V(graph)$gene=="",
                             V(graph)$name,
                             paste(V(graph)$gene,V(graph)$name,sep="\n")),
       vertex.label.cex=1,
       main=title)
}
```


## make graphs for live

get the MR
```{r}
rna_live <- rna[samples_merged$inoc=="SBC_OLD",]
system.time(MR_live <- getcorMR(rna=rna_live, asv=asv_live_small))
```

```{r}
MR_live$cormat[1:10,1:10]
#MR_live$cormat[order(abs(MR_live$cormat[,1]),decreasing = TRUE),1] %>% head(100)
```

Get graphs at different MR.  First set up new tibble to hold them.

```{r}
graphstibble <- tibble(cutoff=c(10, 15, 25) )
graphstibble
```

Make graphs and pull out subgraphs centered on ASVs
```{r}
graphstibble <- graphstibble %>%
  mutate(graph=map(cutoff, ~ get.mr.subgraph(
                      mr.cutoff=.x,
                      mr.matrix=MR_live$MR,
                      annotation = annotation,
                      neighborhood = colnames(asv_live_small)
                    )),
         nodes=map(graph, ~ V(.)$name))
```


## stopped here...modify below....

annotate graph edges with correlation and direction.
```{r}
graphstibble <- graphstibble %>%
  mutate(graph=map(graph, ~ annotateEdges(
                      graph=.x,
                      cormat = MR_live$cormat)
  ))

```


Add useful title
```{r}
graphstibble <- 
graphstibble %>%
  mutate(title=str_c("MR", cutoff, sep="_"))
```

get max and mean correlation to leaf for each graph
```{r}
graphstibble <- graphstibble %>%
  mutate(maxcor=map_dbl(graph, ~ E(.)[.inc("leaf_avg_std")]$cor %>% abs() %>% max() %>% round(3)),
         meancor=map_dbl(graph, ~E(.)[.inc("leaf_avg_std")]$cor %>% abs() %>% mean() %>% round(3)),
         nNode=map_int(graph, ~length(V(.))))
graphstibble %>% select(title, maxcor, meancor, nNode)
```

