---
title: "ASV correlations with WGCNA Eigen genes"
author: "Julin Maloof"
date: "Oct 25, 2024"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


Will focus on live samples only.  Looking for correlations between eigen genes and ASVs.  This is a redo of my previous analysis but using [methods appropriate for the compositional nature of 16S seq data](https://www.frontiersin.org/journals/microbiology/articles/10.3389/fmicb.2017.02224/full)

UPDATE NOV 8: this is proving somewhat challenging.  compositional correlation functions in package propr want count data as input.  So it might make sense to include the (raw) RNAseq data in that and build combined networks.  

Possible approaches:

1. clr data transformation of ASVs and then use the Spearman correlation I used before
2. correlation matrix using package propr, including RNAseq data and then using SpeakEasy2 (or something else) to find networks
3. correlation matrix using spiec-easi, including RNAseq MEs.  Currently I can't get this to work on my Mac if I use the full ASV data set because of errors in huge::huge.  This may be due to extreme sparsity, because if I filter the data to sparsity < 0.95 it works.

Trying #1 below

```{r, eval=FALSE}
install.packages("compositions")
install.packages("zCompositions")
devtools::install_github("zdk123/SpiecEasi")
```

```{r}
library(phyloseq)
#library(compositions) #using SpiecEasi instead
#library(zCompositions)
library(magrittr)
library(tidyverse)
library(knitr)
library(parallel)
library(SpiecEasi)
```

## get ASV (from Scott Klasek)
```{r}
# ps<- readRDS(file="../input/rhizo.ps") # from 2021
ps <- readRDS(file = "../input/sk_rhizo_16s.ps") # from 2024.  
```


```{r}
asv_counts <- otu_table(ps) 
dim(asv_counts)
head(asv_counts[,1:10])
```


```{r}
taxa <- tax_table(ps) # taxonomy data
dim(taxa)
head(taxa)
```


If using 2021 data set use this
```{r, eval=FALSE}
# asv_samples <- sample_data(ps) %>% # sample information
#   as_tibble() %>%
#   mutate(ID=rownames(sample_data(ps)),
#          days_grown=str_pad(days_grown, 2, pad="0"), # get things formatted the same as the RNA names
#          block=str_c("blk", block),
#          inoc={str_replace(inoc, "Autoclaved", "ATM_BLANK") %>%
#              str_replace("Inoculated", "SBC_OLD")},
#          timeofday={
#            str_replace(timeofday, "9:00 AM", "1_morn") %>%
#              str_replace("14:00 PM", "2_afternoon") %>%
#              str_replace("18:00 PM", "3_evening_5.30") %>%
#              str_replace("22:00 PM", "4_night_1") %>%
#              str_replace("4:00 AM", "5_night_2")},
#          sampleID=str_c(inoc,days_grown,timeofday,block, sep = "-")) %>%
#   select("ID", everything())
# dim(asv_samples)
# asv_samples
```

If using 2024 data set use this. 
```{r}
asv_samples <- sample_data(ps) %>% # sample information
  as_tibble() %>%
  mutate(sampling_day=str_pad(sampling_day, 2, pad="0"), # get things formatted the same as the RNA names
         block=str_c("blk", block),
         sampleID=str_c(soil_trt, sampling_day, sampling_time, block, sep = "-")) %>%
  select(ID=name, everything(), -tube_no, -pot, -sampling_day_character, -sampling_day_actual)
dim(asv_samples)
asv_samples
```


```{r}
dups <- asv_samples %>% filter(duplicated(sampleID))  %>% pull(sampleID) 
dups
```


```{r}
asvID_dups <- asv_samples %>% filter(sampleID %in% dups)
asvID_dups
```

```{r}
# not needed for 2024
#sample_data(ps)[asvID_dups$ID,] 
```


## reduce ASV data to live only

```{r}
asv_samples <- asv_samples %>% filter(soil_trt=="SBC_OLD")
asv_counts <- asv_counts[asv_samples$ID,]
dim(asv_counts)
```

## nomalize ASV data

Note: example in help shows a norm_to_total step, but this is unnecessary (I tested it)
```{r}
sparsity <- apply(asv_counts, 2, function(x) sum(x==0)/length(x))
asv_counts_small <- asv_counts[, sparsity < 0.95]
dim(asv_counts_small)

asv_norm <- asv_counts_small %>% clr(mar=1) %>% t()
dim(asv_norm)
```

This works on the subsetted dataset, but not on the full data set.  Will explore later.

```{r, eval=FALSE}
system.time(
asb.cov <- sparseiCov(asv_norm[,1:1000], method = "glasso", verbose = TRUE)
)
```


```{r}
asv_samples <- asv_samples[asv_samples$ID %in% rownames(asv_norm), ]
```


## get modules

```{r}
load("../../RNAseq/output/WGCNA/net.cpm.timecourse.v3.0.log.live.largeCV.Rdata") # not needed?
load("../../RNAseq/output/WGCNA/live.largeCV.RData")
```


### wrangle it

get sample description format to match the ASVs
```{r}
sample.description.timecourse.live <- sample.description.timecourse.live %>%
  mutate(sampleID = {
    str_remove(group, "FPsc-") %>%
      str_c(., "-blk", block)
  }) %>%
  select(-lib.size, -norm.factors, -genotype, -soil_trt, -group.1)
```

### missing samples from ASVs?

```{r}
sample.description.timecourse.live %>% anti_join(asv_samples, by = "sampleID")
```


## get ASV and ME order to match

```{r}
samples_merged <- inner_join(asv_samples, sample.description.timecourse.live, by = "sampleID") %>%
  select(ASV_Sample_ID=ID, RNA_Sample_ID=sample, sampleID, everything()) 

samples_merged
```


```{r}
MEs <- MEs[samples_merged$RNA_Sample_ID,]
head(MEs[,1:10])
asv_norm <- asv_norm[samples_merged$ASV_Sample_ID,]

head(asv_norm[,1:10])
```

Let's give them the same rownames
```{r}
all(rownames(MEs)==samples_merged$RNA_Sample_ID) # double check!
rownames(MEs) <- samples_merged$ASV_Sample_ID
```

## correlations

Note: Kendall and Spearman are essentially the same.  Kendall is slower, removing it.
```{r}

detectCores()

cl <- makeCluster(ceiling(detectCores()*.75))

permute.cor <- function(x, y, method = "pearson", times = 1000) {
  x.perm <- matrix(nrow=length(x), ncol=times)
  for (i in 1:times) x.perm[,i] <- sample(x)
  result <- parApply(cl, x.perm, 2, cor, y, method = method)
}

MEcors <- MEs %>%
  rownames_to_column("ASV_Sample_ID") %>%
  pivot_longer(-ASV_Sample_ID, names_to = "module", values_to = "expression",cols_vary = "slowest") %>%
  nest(.by = module)

#confirm order is still okay

all(MEcors$data[[1]]$ASV_Sample_ID == rownames(asv_norm))

system.time(MEcors <- MEcors %>%
              mutate(pearson.r = map(data, ~ cor(.x$expression, asv_norm)),
                     pearson.perm = map(data, ~ permute.cor(.x$expression, asv_norm, method = "pearson")),
                     spearman.rho = map(data, ~ cor(.x$expression, asv_norm, method = "spearman")),
                     spearman.perm = map(data, ~ permute.cor(.x$expression, asv_norm, method = "spearman"))
                     ))

MEcors <- MEcors %>%
  mutate(pearson.99 = map_dbl(pearson.perm, ~ quantile(abs(.x), 0.99)),
         pearson.max = map_dbl(pearson.r, max),
         pearson.min = map_dbl(pearson.r, min),
         pearson.n.sig.99 = map2_dbl(pearson.r, pearson.99, ~ sum(abs(.x) > .y)),

         spearman.99 = map_dbl(spearman.perm, ~ quantile(abs(.x), 0.99)),
         spearman.max = map_dbl(spearman.rho, max),
         spearman.min = map_dbl(spearman.rho, min),
         spearman.n.sig.99 = map2_dbl(spearman.rho, spearman.99, ~ sum(abs(.x) > .y))
)

stopCluster(cl)

MEcors %>% arrange(desc(pearson.n.sig.99)) %>% select(module, pearson.99:last_col())
```
```{r}
MEcors_clr <- MEcors
asv_counts_small_clr <- asv_counts_small
asv_norm_clr <- asv_norm
samples_merged_clr <- samples_merged
```


```{r}
save(MEcors_clr, asv_counts_small_clr, asv_norm_clr, taxa, samples_merged_clr, file = "../output/ME_ASV_cors_live_clr_normalization.Rdata")
```

