---
title: "09_clustering"
output:
  html_document:
    df_print: paged
editor_options: 
  chunk_output_type: inline
---

Original script by Kazu, updated by Julin

```{r setup}
getwd()
library(knitr)
knitr::opts_chunk$set(echo = TRUE,error=TRUE)
#library(d3heatmap)
library(edgeR);library(tidyverse)
library(readr);library(readxl)
library(stringr);library(glue)
# read functions for expression graph
source("../tools_copy/Expression_pattern_graph.R")
```

get data

```{r}
load("../output/timecourseDGE.Rdata")
cpm.timecourse.v3.0 <- cpm(dge) %>% as_tibble() %>% bind_cols(data.frame(transcript_ID=rownames(dge$counts)),.)
cpm.timecourse.v3.0.log <- cpm(dge,log=TRUE) %>% as_tibble() %>% bind_cols(data.frame(transcript_ID=rownames(dge$counts)),.)

sample.description.timecourse <- dge$samples %>% bind_cols(data.frame(sample=rownames(dge$samples)),.)
```


```{r}
# filtering for selecting only expressed genes
cpm.timecourse.v3.0 <- cpm.timecourse.v3.0[rowSums(cpm.timecourse.v3.0[,-1]>5) >=28,] # 10% of libraries have
dim(cpm.timecourse.v3.0) # [1] 21124   289

# removing low expressed genes from cpm.timecourse.v3.0.log
cpm.timecourse.v3.0.log <- cpm.timecourse.v3.0.log %>% filter(transcript_ID %in% cpm.timecourse.v3.0$transcript_ID)
dim(cpm.timecourse.v3.0.log) #[1] 21124   289
```

# two afternoon data (reading DEG files) (under construction)
```{r, eval=FALSE}
# soil_trt effects (coef=c("SBC_OLD"))
twoafternoon.trtlive.DEGs.all.v3.0anno <- read_csv(file.path("..","output","twoafternoon.trtlive.DEGs.all.v3.0anno.csv"))

# interaction, coef = str_which(colnames(dge.twoafternoon.design),"soil_trtSBC_OLD:")
twoafternoon.interaction.trtlive.samplingday.lrt.DEGs.all.v3.0anno <- read_csv(file=file.path("..","output","twoafternoon.interaction.trtlive.samplingday.lrt.DEGs.all.v3.0anno.csv")) 

# any soil_trt effects, coef = str_which(colnames(dge.twoafternoon.design),"soil_trtSBC_OLD")
twoafternoon.any.trtlive.DEGs.all.v3.0anno <- read_csv(file=file.path("..","output","twoafternoon.any.trtlive.DEGs.all.v3.0anno.csv")) 

```

# diel data (reading DEG files)
```{r, eval=FALSE}
# treatment only
dge.diurnal34.trtlive.DEGs.all.v3.0anno <- read_csv(file=file.path("..","output","dge.diurnal34.trtlive.DEGs.v3.0anno.csv")) # updated
dge.diurnal1314.trtlive.DEGs.all.v3.0anno <- read_csv(file=file.path("..","output","dge.diurnal1314.trtlive.DEGs.all.v3.0anno.csv")) # updated 

# treatment and treatment:time interaction
dge.diurnal34.any.trtlive.DEGs.all.v3.0anno <- read_csv(file=file.path("..","output","dge.diurnal34.any.trtlive.DEGs.all.v3.0anno.csv"))
dge.diurnal1314.any.trtlive.DEGs.all.v3.0anno <- read_csv(file=file.path("..","output","dge.diurnal1314.any.trtlive.DEGs.all.v3.0anno.csv"))

# interaction (coef = str_which(colnames(dge.diurnal34.design),":"))
dge.diurnal34.interaction.trtlive.time.DEGs.all.v3.0anno <- read_csv(file=file.path("..","output","dge.diurnal34.interaction.trtlive.time.DEGs.all.v3.0anno.csv")) # updated
# interaction (coef = str_which(colnames(dge.diurnal1314.design),":"))
dge.diurnal1314.interaction.trtlive.time.DEGs.all.v3.0anno <- read_csv(file=file.path("..","output","dge.diurnal1314.interaction.trtlive.time.DEGs.all.v3.0anno.csv")) # updated

# check
dge.diurnal34.any.trtlive.DEGs.all.v3.0anno %>% filter(FDR<0.05) %>% dim() # [1] 4593   14 

```
# format data

# use log transformed data
```{r}
co.var.df <- function(x) ( 100*apply(x,1,sd)/rowMeans(x) )

cpm.timecourse.v3.0.log$cv <- co.var.df(cpm.timecourse.v3.0.log[,-1])

cpm.timecourse.v3.0.log %>% arrange(desc(cv))

b <- hist(log10(cpm.timecourse.v3.0.log$cv))
b

# use largeCV 
cpm.timecourse.v3.0.log.largeCV <- cpm.timecourse.v3.0.log[abs(cpm.timecourse.v3.0.log$cv) > 10,] 

dim(cpm.timecourse.v3.0.log.largeCV) # [1] 9659  290

hist(log10(cpm.timecourse.v3.0.log.largeCV$cv))

cpm.timecourse.v3.0.log.largeCV <- cpm.timecourse.v3.0.log.largeCV %>%
  select(-cv)
###########

write_csv(cpm.timecourse.v3.0.log.largeCV,path=file.path("..","output","cpm.timecourse.v3.0.log.largeCV.csv.gz"))
```

# WGCNA
## co-expression analysis by WGCNA
```{r eval=FALSE}
# The following setting is important, do not omit.
library(WGCNA) 
options(stringsAsFactors = FALSE)
```

```{r eval=FALSE}
cpm.timecourse.v3.0.log.largeCV <- read_csv(file.path("..","output","cpm.timecourse.v3.0.log.largeCV.csv.gz"))

# 
 datExpr <-t(cpm.timecourse.v3.0.log.largeCV[,-1])
  
 # Choose a set of soft-thresholding powers
  powers = c(c(1:9), seq(from = 10, to=20, by=2))
  sft = pickSoftThreshold(datExpr, powerVector = powers, verbose = 5)  
  
  # Plot the results:
  #sizeGrWindow(9, 5)
  pdf("../output/WGCNA/largeCV.softthresholding.pdf",width=10,height=8)
  par(mfrow = c(1,2));
  cex1 = 0.9;
  # Scale-free topology fit index as a function of the soft-thresholding power
  plot(sft$fitIndices[,1], -sign(sft$fitIndices[,3])*sft$fitIndices[,2],
       xlab="Soft Threshold (power)",ylab="Scale Free Topology Model Fit,signed R^2",type="n",
       main = paste("Scale independence"));
  text(sft$fitIndices[,1], -sign(sft$fitIndices[,3])*sft$fitIndices[,2],
       labels=powers,cex=cex1,col="red");
  # this line corresponds to using an R^2 cut-off of h
  abline(h=0.90,col="red")
  # Mean connectivity as a function of the soft-thresholding power
  plot(sft$fitIndices[,1], sft$fitIndices[,5],
       xlab="Soft Threshold (power)",ylab="Mean Connectivity", type="n",
       main = paste("Mean connectivity"))
  text(sft$fitIndices[,1], sft$fitIndices[,5], labels=powers, cex=cex1,col="red")
  dev.off()
  
  # make modules 
  net = blockwiseModules(datExpr, power = 7,
                         maxBlockSize = 10000,
                         networkType = "signed hybrid",
                         TOMType = "unsigned", minModuleSize = 20,
                         reassignThreshold = 0, mergeCutHeight = 0.25,
                         numericLabels = TRUE, pamRespectsDendro = FALSE,
                         saveTOMs = TRUE,
                         saveTOMFileBase = "cpm.timecourse.v3.0.log.largeCV.TOM",
                         verbose = 3)
  save(net,file="../output/WGCNA/net.cpm.timecourse.v3.0.log.largeCV.Rdata")  
  # open a graphics window
  pdf(file="../output/WGCNA/largeCV.dendrogram.pdf",width=10,height=8)
  # Convert labels to colors for plotting
  mergedColors = labels2colors(net$colors)
  # Plot the dendrogram and the module colors underneath
  plotDendroAndColors(net$dendrograms[[1]], mergedColors[net$blockGenes[[1]]],
                      "Module colors",
                      dendroLabels = FALSE, hang = 0.03,
                      addGuide = TRUE, guideHang = 0.05)
  dev.off()
  # save parameters  
  moduleLabels = net$colors
  moduleColors = labels2colors(net$colors)
  MEs = net$MEs
  geneTree = net$dendrograms[[1]]
  save(MEs, moduleLabels, moduleColors, geneTree,file ="../output/WGCNA/all.largeCV.RData")

```

```{r}
load("../output/WGCNA/all.largeCV.RData")
```


```{r}

# how many modules?
  table(net$colors)

length(table(net$colors)) # 26 modules

```

# adding gene name, annotations
```{r}
cpm.timecourse.v3.0.log.largeCV.modules <- tibble(
  transcript_ID=cpm.timecourse.v3.0.log.largeCV$transcript_ID,
  module=moduleColors
)


## prep
# annotation file for v3.0annotation

Br.v3.0.At.BLAST <- read_csv(file.path("..","Annotation_copy","output","v3.0annotation","Brapa_v3.0_annotated.csv")) 

# This annotation is redundant with name (Br grene). Eg 
Br.v3.0.At.BLAST %>% filter(name=="BraA01g040570.3C")
# reduce the redundancy (112418)
Br.v3.0anno.At.BLAST.highscore <- Br.v3.0.At.BLAST %>% group_by(name) %>% slice_max(score)

cpm.timecourse.v3.0.log.largeCV.modules.annotated <- cpm.timecourse.v3.0.log.largeCV.modules %>%
  left_join(Br.v3.0anno.At.BLAST.highscore, by=c("transcript_ID" = "name")) %>%
  select(AGI, At_symbol, At_full_name, At_short_description, At_Curator_summary, At_Computational_description, perc_ID, aln_length)

```


Gene lengths
```{r, eval=FALSE}
library(Biostrings)
cDNAs <- readDNAStringSet("~/Sequences/ref_genomes/B_rapa/genome/V3.0/Brapa_genome_v3.0_cds.fasta.gz")
CDS.lengths <- nchar(cDNAs)
names(CDS.lengths) <- names(cDNAs)
head(CDS.lengths)
save(CDS.lengths, file="../Annotation_copy/output/v3.0annotation/Brapa_V3.0_CDS_lengths.Rdata")
```

Load lengths and annotations
```{r}
load("../Annotation_copy/output/v3.0annotation/Brgo.v3.0anno.Atgoslim.BP.list.Rdata") # object name `Brgo.v3.0anno.Atgoslim.BP.list`
load("../Annotation_copy/output/v3.0annotation/Brapa_V3.0_CDS_lengths.Rdata") # object name `CDS.lengths`
```

Set up gene list
```{r}
all.expressed.genes.modules <- cpm.timecourse.v3.0.log %>% 
  select(transcript_ID) %>%
  full_join(cpm.timecourse.v3.0.log.largeCV.modules) %>%
  mutate(module=ifelse(is.na(module), "none", module))

all.expressed.genes.modules

CDS.lengths <- CDS.lengths[all.expressed.genes.modules$transcript_ID]

length(CDS.lengths) == nrow(all.expressed.genes.modules)
```

```{r, message=FALSE}
modules.to.gene.vector <- function(module, module.df=all.expressed.genes.modules) {
  module.vector <- as.numeric(module.df$module==module)
  names(module.vector) <- module.df$transcript_ID
  module.vector
}

# Calculate GO enrichment
go.results <- tibble(module=sort(unique(all.expressed.genes.modules$module))) %>%
  mutate(module.gene.vector = map(module, modules.to.gene.vector),
         module.nullp = map(module.gene.vector, ~ nullp(.x, bias.data=CDS.lengths, plot.fit=FALSE)),
         module.goseq = map(module.nullp, ~goseq(.x, 
                                                 gene2cat = Brgo.v3.0anno.Atgoslim.BP.list , 
                                                 test.cats = "GO:BP"))
  )

# Calculate FDR
go.results <- go.results %>%
  mutate(module.goseq = map(module.goseq,
                            ~ mutate(.x, over_represented_FDR = p.adjust(over_represented_pvalue,
                                                                         method = "fdr"))))

# Filter based on FDR
go.results <- go.results %>%
  mutate(module.goseq = map(module.goseq,
                            ~ filter(.x, over_represented_FDR <= 0.05)))

# unnest
go.results <- go.results %>% dplyr::select(module, module.goseq) %>%
  unnest(module.goseq)

write_csv(go.results, "../output/WGCNA/GO_ORA_WGCNA_all_genes.csv")

go.results
```



